internal/pubsub/rabbitpubsub/rabbit.go

```diff
@@ -166,6 +139,54 @@ func (t *topic) SendBatch(ctx context.Context, ms []*driver.Message) error {
 	return <-errc
 }
 
+// Read from the channels established with NotifyPublish and NotifyReturn.
+// Must be called with t.mu held.
+func (t *topic) receiveFromPublishChannels(ctx context.Context, nMessages int) error {
+	// Consume all the acknowledgments for the messages we are publishing.
+	// Since SendBatch (the only caller of this method) holds the lock, we expect
+	// exactly as many acks as messages.
+	var err error
+	nAcks := 0
+	for nAcks < nMessages {
+		select {
+		case <-ctx.Done():
+			// Channel will be in a weird state (not all publish acks consumed, perhaps)
+			// so re-create it next time.
+			t.ch = nil
+			return ctx.Err()
+
+		case ret, ok := <-t.retc:
+			if !ok {
+				// Channel closed. Handled in the pubc case below. But set
+				// the channel to nil to prevent it from being selected again.
+				t.retc = nil
+			} else if err == nil {
+				// The message was returned from the server because it is unroutable.
+				// This will be the error we return, but continue so we drain all
+				// items from pubc. We don't need to re-establish the channel on this
+				// error.
+				err = fmt.Errorf("rabbitpubsub: message returned from %s: %s (code %d)",
+					ret.Exchange, ret.ReplyText, ret.ReplyCode)
+			}
+
+		case conf, ok := <-t.pubc:
+			if !ok {
+				// t.pubc was closed
+				if err == nil {
+					err = errors.New("rabbitpubsub: publish listener closed unexpectedly")
+				}
+				t.ch = nil // re-create the channel on next use
+				return err
+			}
+			nAcks++
+			if !conf.Ack && err == nil {
+				err = errors.New("rabbitpubsub: ack failed on publish")
+			}
+		}
+	}
+	return err
```

Comment:
It might be more valuable to have the error include a count of successes/failures in the batch (e.g., "4 of 4 messages failed to be published, first error: %v"). WDYT?

Reply:
You could even have a wrapped error type that you expose via `ErrorAs` that includes the full slice of errors.

Reply:
#889